scio-core/src/main/java/com/spotify/scio/transforms/BatchDoFn.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * @param <InputT>
+ */
+public class BatchDoFn<InputT> extends DoFn<InputT, Iterable<InputT>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BatchDoFn.class);
+
+    private static final int MAX_LIVE_WINDOWS = 10;
+
+    private static class Buffer<InputT> {
+
+        private final List<InputT> elements;
+
+        private long weight;
+
+        Buffer() {
+            this.elements = new ArrayList<>();
+            this.weight = 0;
+        }
+
+        public List<InputT> getElements() {
+            return elements;
+        }
+
+        public long getWeight() {
+            return weight;
+        }
+
+        public void add(InputT element, long weight) {
+            elements.add(element);
+            this.weight += weight;
+        }
+    }
+
+    private final long maxWeight;
+    private final SerializableFunction<InputT, Long> weigher;
+    private transient Map<BoundedWindow, Buffer<InputT>> buffers;
+
+    public BatchDoFn(
+            long maxWeight,
+            SerializableFunction<InputT, Long> weigher
+    ) {
+        this.maxWeight = maxWeight;
+        this.weigher = weigher;
+    }
+
+    @Setup
+    public void setup() {
+        this.buffers = new HashMap<>();
+    }
+
+    @ProcessElement
+    public void processElement(
+            @Element InputT element,
+            BoundedWindow window,
+            OutputReceiver<Iterable<InputT>> receiver
+    ) {
+        LOG.debug("*** BATCH *** Add element for window {} ", window);
+        Buffer<InputT> buffer = buffers.computeIfAbsent(window, w -> new Buffer<>());
+        long weight = weigher.apply(element);
+        buffer.add(element, weight);
+
+        if (buffer.getWeight() >= maxWeight) {
+            LOG.debug("*** END OF BATCH *** for window {}", window);
+            flushBatch(window, buffer, receiver);
+        } else if (buffers.size() > MAX_LIVE_WINDOWS) {
+            // flush the biggest buffer if we get too many parallel windows
+            Map.Entry<BoundedWindow, Buffer<InputT>> discarded =
+                    Collections.max(buffers.entrySet(), Comparator.comparingLong(o -> o.getValue().getWeight()));
+            BoundedWindow discardedWindow = discarded.getKey();
+            Buffer<InputT> discardedBatch = discarded.getValue();
+            LOG.debug("*** END OF BATCH *** for window {}", discardedWindow);
+            flushBatch(discardedWindow, discardedBatch, receiver);
+        }
+    }
+
+    @FinishBundle
+    public void finishBundle(FinishBundleContext context) {
+        for (Map.Entry<BoundedWindow, Buffer<InputT>> wb : buffers.entrySet()) {
+            BoundedWindow window = wb.getKey();
+            Buffer<InputT> buffer = wb.getValue();
+            context.output(buffer.getElements(), window.maxTimestamp(), window);
+        }
+        buffers.clear();
+    }
+
+    private void flushBatch(
+            BoundedWindow window,
+            Buffer<InputT> buffer,
```

Comment:
isn't it less error prone to only pass the key, and do the look up inside flush method?

Reply:
is it to avoid one extra look up?

Reply:
yes, `remove` returns the value, we can do with the key only